README.md
# export-regex

[![github release](https://img.shields.io/github/v/release/flex-development/export-regex.svg?include_prereleases&sort=semver)](https://github.com/flex-development/export-regex/releases/latest)
[![npm](https://img.shields.io/npm/v/@flex-development/export-regex.svg)](https://npmjs.com/package/@flex-development/export-regex)
[![codecov](https://codecov.io/github/flex-development/export-regex/branch/main/graph/badge.svg?token=f79rIEL1sK)](https://codecov.io/github/flex-development/export-regex)
[![module type: esm](https://img.shields.io/badge/module%20type-esm-brightgreen)](https://github.com/voxpelli/badges-cjs-esm)
[![license](https://img.shields.io/github/license/flex-development/export-regex.svg)](LICENSE.md)
[![conventional commits](https://img.shields.io/badge/-conventional%20commits-fe5196?logo=conventional-commits&logoColor=ffffff)](https://conventionalcommits.org/)
[![typescript](https://img.shields.io/badge/-typescript-3178c6?logo=typescript&logoColor=ffffff)](https://typescriptlang.org/)
[![vitest](https://img.shields.io/badge/-vitest-6e9f18?style=flat&logo=vitest&logoColor=ffffff)](https://vitest.dev/)
[![yarn](https://img.shields.io/badge/-yarn-2c8ebb?style=flat&logo=yarn&logoColor=ffffff)](https://yarnpkg.com/)

`export` statement regex.

## Contents

- [What is this?](#what-is-this)
- [When should I use this?](#when-should-i-use-this)
- [Install](#install)
- [Use](#use)
- [API](#api)
  - [`EXPORT_AGGREGATE_REGEX`](#export_aggregate_regex)
  - [`EXPORT_DECLARATION_REGEX`](#export_declaration_regex)
  - [`EXPORT_DEFAULT_REGEX`](#export_default_regex)
  - [`EXPORT_LIST_REGEX`](#export_list_regex)
- [Types](#types)
- [Related](#related)
- [Contribute](#contribute)

## What is this?

This package contains regular expressions for matching [`export`][1] statements.

## When should I use this?

Use this package when you need to match `export` statements.

**Note**:

- Statements in docblock (`/** */`), multiline (`/* */`), and single-line (`//`) comments are ignored
- Expressions are ECMAScript-compatible. They have **not** been tested with other flavors (PCRE, PCRE2, etc)

## Install

This package is [ESM only][2].

```sh
yarn add @flex-development/export-regex
```

From Git:

```sh
yarn add @flex-development/export-regex@flex-development/export-regex
```

<blockquote>
  <small>
    See <a href='https://yarnpkg.com/features/protocols#git'>Git - Protocols | Yarn</a>
    &nbsp;for details on requesting a specific branch, commit, or tag.
  </small>
</blockquote>

## Use

Suppose we have the following module:

```ts
import * as regexp from '@flex-development/export-regex'
import { at, omit, select, type Times } from '@flex-development/tutils'
import { dedent } from 'ts-dedent'

const code: Times<4, string> = [
  dedent`
    export { defineBuildConfig, type BuildConfig } from "#src"
    export type {
      JsonObject,
      LiteralUnion,
      Nullable
    } from '@flex-development/tutils'
    export * as constants from "./constants"
    export * from './interfaces'
  `,
  dedent`
    export const ESM_SYNTAX_REGEX: RegExp = /pattern/gm;
    export declare const RESOLVE_EXTENSIONS: string[]
    export enum Snack{}
    export interface User {}
    export class House {}
    export declare abstract class House extends Building {}
    export abstract class House {}
    export async function run() {}
    export function* foo() {}
    export namespace Foo {}
    export type Foo = 'foo'
    export function functionName() {}
    export class ClassName {}
    export const { name1, name2: bar } = o;
    export const [ name1, name2 ] = array;
  `,
  dedent`
    export default function () {}
    export default async function() {}
    export default class {}

    export default function functionName() {}
    export default function* generatorName() {}
    export default class ClassName {}
    export default abstract class ClassName {}

    export default async () => {}
    export default () => {}
    export default arg => {}

    export default foo
    export default 1 + 1;
  `,
  dedent`
    export { defineBuildConfig, type BuildConfig }
    export type {
      JsonObject,
      LiteralUnion,
      Nullable
    }
  `
]

const print = (matches: IterableIterator<RegExpMatchArray>): void => {
  console.debug(select([...matches], null, match => omit(match, ['input'])))
}

print(at(code, 0).matchAll(regexp.EXPORT_AGGREGATE_REGEX))
print(at(code, 1).matchAll(regexp.EXPORT_DECLARATION_REGEX))
print(at(code, 2).matchAll(regexp.EXPORT_DEFAULT_REGEX))
print(at(code, 3).matchAll(regexp.EXPORT_LIST_REGEX))
```

...running that yields:

```zsh
[
  {
    '0': 'export { defineBuildConfig, type BuildConfig } from "#src"',
    '1': undefined,
    '2': '{ defineBuildConfig, type BuildConfig }',
    '3': '#src',
    index: 0,
    groups: [Object: null prototype] {
      type: undefined,
      exports: '{ defineBuildConfig, type BuildConfig }',
      specifier: '#src'
    }
  },
  {
    '0': 'export type {\n' +
      '  JsonObject,\n' +
      '  LiteralUnion,\n' +
      '  Nullable\n' +
      "} from '@flex-development/tutils'",
    '1': 'type',
    '2': '{\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}',
    '3': '@flex-development/tutils',
    index: 59,
    groups: [Object: null prototype] {
      type: 'type',
      exports: '{\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}',
      specifier: '@flex-development/tutils'
    }
  },
  {
    '0': 'export * as constants from "./constants"',
    '1': undefined,
    '2': '* as constants',
    '3': './constants',
    index: 148,
    groups: [Object: null prototype] {
      type: undefined,
      exports: '* as constants',
      specifier: './constants'
    }
  },
  {
    '0': "export * from './interfaces'",
    '1': undefined,
    '2': '*',
    '3': './interfaces',
    index: 189,
    groups: [Object: null prototype] {
      type: undefined,
      exports: '*',
      specifier: './interfaces'
    }
  }
]
[
  {
    '0': 'export const ESM_SYNTAX_REGEX',
    '1': undefined,
    '2': 'const',
    '3': 'ESM_SYNTAX_REGEX',
    index: 0,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'const',
      exports: 'ESM_SYNTAX_REGEX'
    }
  },
  {
    '0': 'export declare const RESOLVE_EXTENSIONS',
    '1': 'declare',
    '2': 'const',
    '3': 'RESOLVE_EXTENSIONS',
    index: 53,
    groups: [Object: null prototype] {
      modifiers: 'declare',
      declaration: 'const',
      exports: 'RESOLVE_EXTENSIONS'
    }
  },
  {
    '0': 'export enum Snack',
    '1': undefined,
    '2': 'enum',
    '3': 'Snack',
    index: 103,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'enum',
      exports: 'Snack'
    }
  },
  {
    '0': 'export interface User',
    '1': undefined,
    '2': 'interface',
    '3': 'User',
    index: 123,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'interface',
      exports: 'User'
    }
  },
  {
    '0': 'export class House',
    '1': undefined,
    '2': 'class',
    '3': 'House',
    index: 148,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'class',
      exports: 'House'
    }
  },
  {
    '0': 'export declare abstract class House',
    '1': 'declare abstract',
    '2': 'class',
    '3': 'House',
    index: 170,
    groups: [Object: null prototype] {
      modifiers: 'declare abstract',
      declaration: 'class',
      exports: 'House'
    }
  },
  {
    '0': 'export abstract class House',
    '1': 'abstract',
    '2': 'class',
    '3': 'House',
    index: 226,
    groups: [Object: null prototype] {
      modifiers: 'abstract',
      declaration: 'class',
      exports: 'House'
    }
  },
  {
    '0': 'export async function run',
    '1': 'async',
    '2': 'function',
    '3': 'run',
    index: 257,
    groups: [Object: null prototype] {
      modifiers: 'async',
      declaration: 'function',
      exports: 'run'
    }
  },
  {
    '0': 'export function* foo',
    '1': undefined,
    '2': 'function*',
    '3': 'foo',
    index: 288,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'function*',
      exports: 'foo'
    }
  },
  {
    '0': 'export namespace Foo',
    '1': undefined,
    '2': 'namespace',
    '3': 'Foo',
    index: 314,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'namespace',
      exports: 'Foo'
    }
  },
  {
    '0': 'export type Foo',
    '1': undefined,
    '2': 'type',
    '3': 'Foo',
    index: 338,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'type',
      exports: 'Foo'
    }
  },
  {
    '0': 'export function functionName',
    '1': undefined,
    '2': 'function',
    '3': 'functionName',
    index: 362,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'function',
      exports: 'functionName'
    }
  },
  {
    '0': 'export class ClassName',
    '1': undefined,
    '2': 'class',
    '3': 'ClassName',
    index: 396,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'class',
      exports: 'ClassName'
    }
  },
  {
    '0': 'export const { name1, name2: bar } = o',
    '1': undefined,
    '2': 'const',
    '3': '{ name1, name2: bar }',
    index: 422,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'const',
      exports: '{ name1, name2: bar }'
    }
  },
  {
    '0': 'export const [ name1, name2 ] = array',
    '1': undefined,
    '2': 'const',
    '3': '[ name1, name2 ]',
    index: 462,
    groups: [Object: null prototype] {
      modifiers: undefined,
      declaration: 'const',
      exports: '[ name1, name2 ]'
    }
  }
]
[
  {
    '0': 'export default function',
    '1': undefined,
    '2': 'function',
    '3': undefined,
    index: 0,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: 'function',
      exports: undefined
    }
  },
  {
    '0': 'export default async function',
    '1': 'async',
    '2': 'function',
    '3': undefined,
    index: 30,
    groups: [Object: null prototype] {
      modifiers: 'async',
      kind: 'function',
      exports: undefined
    }
  },
  {
    '0': 'export default class',
    '1': undefined,
    '2': 'class',
    '3': undefined,
    index: 65,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: 'class',
      exports: undefined
    }
  },
  {
    '0': 'export default function functionName',
    '1': undefined,
    '2': 'function',
    '3': 'functionName',
    index: 90,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: 'function',
      exports: 'functionName'
    }
  },
  {
    '0': 'export default function* generatorName',
    '1': undefined,
    '2': 'function*',
    '3': 'generatorName',
    index: 132,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: 'function*',
      exports: 'generatorName'
    }
  },
  {
    '0': 'export default class ClassName',
    '1': undefined,
    '2': 'class',
    '3': 'ClassName',
    index: 176,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: 'class',
      exports: 'ClassName'
    }
  },
  {
    '0': 'export default abstract class ClassName',
    '1': 'abstract',
    '2': 'class',
    '3': 'ClassName',
    index: 210,
    groups: [Object: null prototype] {
      modifiers: 'abstract',
      kind: 'class',
      exports: 'ClassName'
    }
  },
  {
    '0': 'export default async',
    '1': 'async',
    '2': undefined,
    '3': undefined,
    index: 254,
    groups: [Object: null prototype] {
      modifiers: 'async',
      kind: undefined,
      exports: undefined
    }
  },
  {
    '0': 'export default',
    '1': undefined,
    '2': undefined,
    '3': undefined,
    index: 284,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: undefined,
      exports: undefined
    }
  },
  {
    '0': 'export default',
    '1': undefined,
    '2': undefined,
    '3': undefined,
    index: 308,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: undefined,
      exports: undefined
    }
  },
  {
    '0': 'export default foo',
    '1': undefined,
    '2': undefined,
    '3': 'foo',
    index: 334,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: undefined,
      exports: 'foo'
    }
  },
  {
    '0': 'export default',
    '1': undefined,
    '2': undefined,
    '3': undefined,
    index: 353,
    groups: [Object: null prototype] {
      modifiers: undefined,
      kind: undefined,
      exports: undefined
    }
  }
]
[
  {
    '0': 'export { defineBuildConfig, type BuildConfig }',
    '1': undefined,
    '2': '{ defineBuildConfig, type BuildConfig }',
    index: 0,
    groups: [Object: null prototype] {
      type: undefined,
      exports: '{ defineBuildConfig, type BuildConfig }'
    }
  },
  {
    '0': 'export type {\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}',
    '1': 'type',
    '2': '{\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}',
    index: 47,
    groups: [Object: null prototype] {
      type: 'type',
      exports: '{\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}'
    }
  }
]
```

## API

This package exports the following identifiers:

- [`EXPORT_AGGREGATE_REGEX`](#export_aggregate_regex)
- [`EXPORT_DECLARATION_REGEX`](#export_declaration_regex)
- [`EXPORT_DEFAULT_REGEX`](#export_default_regex)
- [`EXPORT_LIST_REGEX`](#export_list_regex)

There is no default export.

### `EXPORT_AGGREGATE_REGEX`

- **Source**: [`src/export-aggregate.ts`](src/export-aggregate.ts)

Aggregate `export` statement regex. Ignores matches in comments.

### `EXPORT_DECLARATION_REGEX`

- **Source**: [`src/export-declaration.ts`](src/export-declaration.ts)

Declaration `export` statement regex. Ignores matches in comments.

**Requires unicode support ([flag u][3])**.

### `EXPORT_DEFAULT_REGEX`

- **Source**: [`src/export-default.ts`](src/export-default.ts)

Default `export` statement regex. Ignores matches in comments.

Declaration bodies and expressions that are not identifiers are not captured.

**Requires unicode support ([flag u][3])**.

### `EXPORT_LIST_REGEX`

- **Source**: [`src/export-list.ts`](src/export-list.ts)

List `export` statement regex. Ignores matches in comments.

## Types

This package is fully typed with [TypeScript][4].

## Related

- [`import-regex`][5] &mdash; `import` statement regex

## Contribute

See [`CONTRIBUTING.md`](CONTRIBUTING.md).

[1]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/export
[2]: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c
[3]: https://javascript.info/regexp-unicode
[4]: https://www.typescriptlang.org
[5]: https://github.com/flex-development/import-regex


CHANGELOG.md
## [2.0.0](https://github.com/flex-development/export-regex/compare/1.0.2...2.0.0) (2023-08-01)


### ⚠ BREAKING CHANGES

* **nvm:** require `>=16.20.0`

### :package: Build

* **deps-dev:** Bump @commitlint/cli from 17.5.1 to 17.6.1 ([#150](https://github.com/flex-development/export-regex/issues/150)) ([ac8e136](https://github.com/flex-development/export-regex/commit/ac8e136f19c2ef36c43a00799391f298dc0db4aa))
* **deps-dev:** Bump @commitlint/cli from 17.6.1 to 17.6.3 ([#178](https://github.com/flex-development/export-regex/issues/178)) ([9d7a929](https://github.com/flex-development/export-regex/commit/9d7a929228d11a4bd2793a3b5f271dbffa5d9564))
* **deps-dev:** Bump @flex-development/esm-types from 1.0.0 to 2.0.0 ([#242](https://github.com/flex-development/export-regex/issues/242)) ([0290f33](https://github.com/flex-development/export-regex/commit/0290f33b690002fead0f20c826084ee505974d02))
* **deps-dev:** Bump @flex-development/mlly from 1.0.0-alpha.13 to 1.0.0-alpha.15 ([#92](https://github.com/flex-development/export-regex/issues/92)) ([6bfe35e](https://github.com/flex-development/export-regex/commit/6bfe35ebb07173a625e27c2c0461b859d4fa0cfb))
* **deps-dev:** Bump @flex-development/pathe from 1.0.3 to 2.0.0 ([#243](https://github.com/flex-development/export-regex/issues/243)) ([2c17225](https://github.com/flex-development/export-regex/commit/2c17225929a1a06791ef9480cbdd7ccabf3de797))
* **deps-dev:** Bump @typescript-eslint/eslint-plugin from 5.58.0 to 5.59.0 ([#155](https://github.com/flex-development/export-regex/issues/155)) ([016f510](https://github.com/flex-development/export-regex/commit/016f5105bf07565012a5282c9ded633f700bb3e9))
* **deps-dev:** Bump @typescript-eslint/eslint-plugin from 5.59.0 to 5.59.1 ([#170](https://github.com/flex-development/export-regex/issues/170)) ([b97c64e](https://github.com/flex-development/export-regex/commit/b97c64e55d5d2003a36a8267852dc0e9523ee03d))
* **deps-dev:** Bump @typescript-eslint/eslint-plugin from 5.59.1 to 5.59.2 ([#174](https://github.com/flex-development/export-regex/issues/174)) ([2111aef](https://github.com/flex-development/export-regex/commit/2111aef378728e99868d3f7b6aa3dfcb87e2dc95))
* **deps-dev:** Bump @typescript-eslint/eslint-plugin from 5.59.2 to 5.59.6 ([#191](https://github.com/flex-development/export-regex/issues/191)) ([dc38f58](https://github.com/flex-development/export-regex/commit/dc38f585d6b69224f9739ef98475ded2e0c052f5))
* **deps-dev:** Bump @typescript-eslint/parser from 5.54.0 to 5.54.1 ([#98](https://github.com/flex-development/export-regex/issues/98)) ([e72a5dd](https://github.com/flex-development/export-regex/commit/e72a5dd13660c6c92d8809c0f8c6f21af4e69fcb))
* **deps-dev:** Bump @typescript-eslint/parser from 5.58.0 to 5.59.0 ([#157](https://github.com/flex-development/export-regex/issues/157)) ([33c4cde](https://github.com/flex-development/export-regex/commit/33c4cdee50482d9264b6ffb701885984a899a00d))
* **deps-dev:** Bump @typescript-eslint/parser from 5.59.0 to 5.59.1 ([#168](https://github.com/flex-development/export-regex/issues/168)) ([a95a209](https://github.com/flex-development/export-regex/commit/a95a209e11b23f9b52fa92683432d037b8879584))
* **deps-dev:** Bump @typescript-eslint/parser from 5.59.1 to 5.59.2 ([#173](https://github.com/flex-development/export-regex/issues/173)) ([2bc7ca6](https://github.com/flex-development/export-regex/commit/2bc7ca6d92bb13d8418d8a62d688f34be7687910))
* **deps-dev:** Bump @vitest/coverage-c8 from 0.30.0 to 0.30.1 ([#144](https://github.com/flex-development/export-regex/issues/144)) ([eb372d4](https://github.com/flex-development/export-regex/commit/eb372d4a7936e78d24d7b4af9ebb5da16f8bcb2c))
* **deps-dev:** Bump @vitest/coverage-c8 from 0.30.1 to 0.31.0 ([#176](https://github.com/flex-development/export-regex/issues/176)) ([42c1019](https://github.com/flex-development/export-regex/commit/42c1019e51671422ac6810cd430fb128c874e891))
* **deps-dev:** Bump @vitest/coverage-c8 from 0.31.0 to 0.31.1 ([#193](https://github.com/flex-development/export-regex/issues/193)) ([ee7e207](https://github.com/flex-development/export-regex/commit/ee7e207df30ba465bf2e40f2e590619bfa9780c5))
* **deps-dev:** Bump @vitest/coverage-c8 from 0.31.1 to 0.31.4 ([#213](https://github.com/flex-development/export-regex/issues/213)) ([f40bc30](https://github.com/flex-development/export-regex/commit/f40bc303d2a5c44b1cfd9969a4f38e7779a83410))
* **deps-dev:** Bump @vitest/coverage-c8 from 0.31.4 to 0.32.0 ([#214](https://github.com/flex-development/export-regex/issues/214)) ([2b37eea](https://github.com/flex-development/export-regex/commit/2b37eea3ec272dc4f8081d7a6929bf3a7d11d8ca))
* **deps-dev:** Bump @vitest/ui from 0.30.0 to 0.30.1 ([#143](https://github.com/flex-development/export-regex/issues/143)) ([1a8e117](https://github.com/flex-development/export-regex/commit/1a8e117a808a864c61b08f447fc90d137b8911d5))
* **deps-dev:** Bump @vitest/ui from 0.30.1 to 0.31.0 ([#179](https://github.com/flex-development/export-regex/issues/179)) ([88fcecf](https://github.com/flex-development/export-regex/commit/88fcecf0ac14a66764a68bd279dc025e42604649))
* **deps-dev:** Bump @vitest/ui from 0.31.0 to 0.32.0 ([#217](https://github.com/flex-development/export-regex/issues/217)) ([9480e2d](https://github.com/flex-development/export-regex/commit/9480e2d951d8c5bb54f6e0040e2be68f8abb7820))
* **deps-dev:** Bump cspell from 6.27.0 to 6.30.2 ([#114](https://github.com/flex-development/export-regex/issues/114)) ([2e47029](https://github.com/flex-development/export-regex/commit/2e4702962f2b9fc8c88ddb333aa710a5cc3b0f36))
* **deps-dev:** Bump esbuild from 0.17.16 to 0.17.17 ([#152](https://github.com/flex-development/export-regex/issues/152)) ([7fb980d](https://github.com/flex-development/export-regex/commit/7fb980dfaed1409711d9c44e2761823f4236b9d4))

src/__tests__/export-list.spec-d.ts
/**
 * @file Type Tests - EXPORT_LIST_REGEX
 * @module export-regex/tests/unit-d/list
 */

import type TEST_SUBJECT from '../export-list'

describe('unit-d:EXPORT_LIST_REGEX', () => {
  it('should be instance of RegExp', () => {
    expectTypeOf<typeof TEST_SUBJECT>().toEqualTypeOf<RegExp>
  })
})


vitest-env.d.ts
/// <reference types='vitest/globals' />

interface ImportMetaEnv extends import('vitest/importMeta').ImportMetaEnv {
  readonly BASE_URL: string
  readonly CI: 'false' | 'true'
  readonly DEV: '1' | import('@flex-development/tutils').EmptyString
  readonly MODE: import('@flex-development/tutils').NodeEnv.TEST
  readonly NODE_ENV: import('@flex-development/tutils').NodeEnv.TEST
  readonly NODE_OPTIONS: string
  readonly PROD: '1' | import('@flex-development/tutils').EmptyString
  readonly PWD: string
  readonly SSR: '1' | import('@flex-development/tutils').EmptyString
  readonly TEST: 'true'
  readonly USER: string
  readonly VITEST: 'true'
  readonly VITEST_CLI_WRAPPER: 'true'
  readonly VITEST_MODE: 'DEV' | 'RUN'
  readonly VITEST_POOL_ID: `${number}`
  readonly VITEST_WORKER_ID: `${number}`
  readonly VITE_ROOT: string
  readonly VITE_USER_NODE_ENV: import('@flex-development/tutils').NodeEnv.TEST
}

interface ImportMeta {
  readonly env: ImportMetaEnv
}


src/export-declaration.ts
/**
 * @file EXPORT_DECLARATION_REGEX
 * @module export-regex/declaration
 */

/**
 * Declaration `export` statement regex. Ignores matches in comments.
 *
 * **Note**: Requires unicode support ([flag `u`][1]).
 *
 * [1]: https://javascript.info/regexp-unicode
 *
 * @see https://regex101.com/r/8HpMrA
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/export#syntax
 *
 * @example
 *  import { EXPORT_DECLARATION_REGEX } from '@flex-development/export-regex'
 *  import { dedent } from 'ts-dedent'
 *
 *  const code: string = dedent`
 *    export const ESM_SYNTAX_REGEX: RegExp = /pattern/gm;
 *    export declare const RESOLVE_EXTENSIONS: string[]
 *    export enum Snack{}
 *    export const enum Snack {}
 *    export interface User {}
 *    export class House {}
 *    export declare abstract class House extends Building {}
 *    export abstract class House {}
 *    export async function run() {}
 *    export async function* run() {}
 *    export function foo() {}
 *    export function* foo() {}
 *    export const $foo = 42
 *    export const $foo= 42
 *    export namespace Foo {}
 *    export type Foo = 'foo'
 *    export let name1, name2; // also var
 *    export const name1 = 1, name2 = 2; // also var, let
 *    export function functionName() {}
 *    export class ClassName {}
 *    export function* generatorFunctionName() {}
 *    export const { name1, name2: bar } = o;
 *    export const [ name1, name2 ] = array;
 *  `
 *
 *  const print = (matches: IterableIterator<RegExpMatchArray>): void => {
 *    console.debug([...matches].map(match => omit(match, ['input'])))
 *  }
 *
 *  print(code.matchAll(EXPORT_DECLARATION_REGEX))
 *  // [
 *  //   {
 *  //     '0': 'export const ESM_SYNTAX_REGEX',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': 'ESM_SYNTAX_REGEX',
 *  //     index: 0,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: 'ESM_SYNTAX_REGEX'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export declare const RESOLVE_EXTENSIONS',
 *  //     '1': 'declare',
 *  //     '2': 'const',
 *  //     '3': 'RESOLVE_EXTENSIONS',
 *  //     index: 53,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'declare',
 *  //       declaration: 'const',
 *  //       exports: 'RESOLVE_EXTENSIONS'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export enum Snack',
 *  //     '1': undefined,
 *  //     '2': 'enum',
 *  //     '3': 'Snack',
 *  //     index: 103,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'enum',
 *  //       exports: 'Snack'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const enum Snack',
 *  //     '1': undefined,
 *  //     '2': 'const enum',
 *  //     '3': 'Snack',
 *  //     index: 123,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const enum',
 *  //       exports: 'Snack'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export interface User',
 *  //     '1': undefined,
 *  //     '2': 'interface',
 *  //     '3': 'User',
 *  //     index: 150,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'interface',
 *  //       exports: 'User'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export class House',
 *  //     '1': undefined,
 *  //     '2': 'class',
 *  //     '3': 'House',
 *  //     index: 175,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'class',
 *  //       exports: 'House'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export declare abstract class House',
 *  //     '1': 'declare abstract',
 *  //     '2': 'class',
 *  //     '3': 'House',
 *  //     index: 197,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'declare abstract',
 *  //       declaration: 'class',
 *  //       exports: 'House'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export abstract class House',
 *  //     '1': 'abstract',
 *  //     '2': 'class',
 *  //     '3': 'House',
 *  //     index: 253,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'abstract',
 *  //       declaration: 'class',
 *  //       exports: 'House'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export async function run',
 *  //     '1': 'async',
 *  //     '2': 'function',
 *  //     '3': 'run',
 *  //     index: 284,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'async',
 *  //       declaration: 'function',
 *  //       exports: 'run'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export async function* run',
 *  //     '1': 'async',
 *  //     '2': 'function*',
 *  //     '3': 'run',
 *  //     index: 315,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'async',
 *  //       declaration: 'function*',
 *  //       exports: 'run'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function foo',
 *  //     '1': undefined,
 *  //     '2': 'function',
 *  //     '3': 'foo',
 *  //     index: 347,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function',
 *  //       exports: 'foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function* foo',
 *  //     '1': undefined,
 *  //     '2': 'function*',
 *  //     '3': 'foo',
 *  //     index: 372,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function*',
 *  //       exports: 'foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const $foo',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '$foo',
 *  //     index: 398,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '$foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const $foo',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '$foo',
 *  //     index: 421,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '$foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export namespace Foo',
 *  //     '1': undefined,
 *  //     '2': 'namespace',
 *  //     '3': 'Foo',
 *  //     index: 443,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'namespace',
 *  //       exports: 'Foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export type Foo',
 *  //     '1': undefined,
 *  //     '2': 'type',
 *  //     '3': 'Foo',
 *  //     index: 467,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'type',
 *  //       exports: 'Foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export let name1, name2',
 *  //     '1': undefined,
 *  //     '2': 'let',
 *  //     '3': 'name1, name2',
 *  //     index: 491,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'let',
 *  //       exports: 'name1, name2'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const name1 = 1, name2 = 2',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': 'name1 = 1, name2 = 2',
 *  //     index: 536,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: 'name1 = 1, name2 = 2'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function functionName',
 *  //     '1': undefined,
 *  //     '2': 'function',
 *  //     '3': 'functionName',
 *  //     index: 596,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function',
 *  //       exports: 'functionName'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export class ClassName',
 *  //     '1': undefined,
 *  //     '2': 'class',
 *  //     '3': 'ClassName',
 *  //     index: 639,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'class',
 *  //       exports: 'ClassName'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function* generatorFunctionName',
 *  //     '1': undefined,
 *  //     '2': 'function*',
 *  //     '3': 'generatorFunctionName',
 *  //     index: 674,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function*',
 *  //       exports: 'generatorFunctionName'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const { name1, name2: bar } = o',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '{ name1, name2: bar }',
 *  //     index: 727,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '{ name1, name2: bar }'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const [ name1, name2 ] = array',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '[ name1, name2 ]',
 *  //     index: 767,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '[ name1, name2 ]'
 *  //     }
 *  //   }
 *  // ]
 *
 * @const {RegExp} EXPORT_DECLARATION_REGEX
 */
const EXPORT_DECLARATION_REGEX: RegExp =
  /(?<=^|[\n;](?:[\t ]*(?:\w+ )?)?)export\s*(?<modifiers>(?:\s*declare|\s*abstract|\s*async)+)?\s*(?<declaration>class|const +enum|const|enum|function\*?|interface|let|namespace|type(?! *\{)|var)\s+(?<exports>(?:[$_\p{ID_Start}][$\u200C\u200D\p{ID_Continue}]*(?=[\s=:;/({])(?!.*?,))|(?:[\w\t\n\r .,:$'"=-]+)|(?:[{[][\w\t\n\r .,:$'"-]+[}\]]))(?:\s*=\s*[$_\p{ID_Start}][$\u200C\u200D\p{ID_Continue}]*)?/gu

export default EXPORT_DECLARATION_REGEX
